File: hyperlift--backend/src/main/java/com/hyperlift/service/ExerciseService.java
package com.hyperlift.service;

import com.hyperlift.dao.ExerciseDao;
import com.hyperlift.dto.ExerciseRequest;
import com.hyperlift.dto.ExerciseResponse;
import com.hyperlift.entity.Exercise;
import com.hyperlift.exception.ExerciseNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional
public class ExerciseService {

    private final ExerciseDao exerciseDao;

    public ExerciseService(ExerciseDao exerciseDao) {
        this.exerciseDao = exerciseDao;
    }

    public List<ExerciseResponse> getAllExercises() {
        List<Exercise> exercises = exerciseDao.findAll();
        return exercises.stream()
                .map(this::convertToExerciseResponse)
                .collect(Collectors.toList());
    }

    public ExerciseResponse getExerciseById(Long id) {
        Exercise exercise = exerciseDao.findById(id)
                .orElseThrow(() -> new ExerciseNotFoundException(id));
        return convertToExerciseResponse(exercise);
    }

    public List<ExerciseResponse> getExercisesByMuscleGroup(String muscleGroup) {
        List<Exercise> exercises = exerciseDao.findByMuscleGroup(muscleGroup);
        return exercises.stream()
                .map(this::convertToExerciseResponse)
                .collect(Collectors.toList());
    }

    public List<ExerciseResponse> searchExercises(String name) {
        List<Exercise> exercises = exerciseDao.findByNameContainingIgnoreCase(name);
        return exercises.stream()
                .map(this::convertToExerciseResponse)
                .collect(Collectors.toList());
    }

    public ExerciseResponse createExercise(ExerciseRequest request) {

        if (exerciseDao.existsByNameIgnoreCase(request.getName())) {
            throw new IllegalArgumentException("Exercise already exists with name: " + request.getName());
        }


        Exercise exercise = Exercise.builder()
                .name(request.getName())
                .description(request.getDescription())
                .muscleGroup(request.getMuscleGroup())
                .equipment(request.getEquipment())
                .difficultyLevel(request.getDifficultyLevel())
                .videoUrl(request.getVideoUrl())
                .build();

        Exercise savedExercise = exerciseDao.save(exercise);
        return convertToExerciseResponse(savedExercise);
    }

    public ExerciseResponse updateExercise(Long id, ExerciseRequest request) {
        Exercise exercise = exerciseDao.findById(id)
                .orElseThrow(() -> new ExerciseNotFoundException(id));

        exercise.setName(request.getName());
        exercise.setDescription(request.getDescription());
        exercise.setMuscleGroup(request.getMuscleGroup());
        exercise.setEquipment(request.getEquipment());
        exercise.setDifficultyLevel(request.getDifficultyLevel());
        exercise.setVideoUrl(request.getVideoUrl());

        Exercise updatedExercise = exerciseDao.save(exercise);
        return convertToExerciseResponse(updatedExercise);
    }

    public void deleteExercise(Long id) {
        if (!exerciseDao.existsById(id)) {
            throw new ExerciseNotFoundException(id);
        }
        exerciseDao.deleteById(id);
    }

    private ExerciseResponse convertToExerciseResponse(Exercise exercise) {
        return ExerciseResponse.builder()
                .id(exercise.getId())
                .name(exercise.getName())
                .description(exercise.getDescription())
                .muscleGroup(exercise.getMuscleGroup())
                .equipment(exercise.getEquipment())
                .difficultyLevel(exercise.getDifficultyLevel())
                .videoUrl(exercise.getVideoUrl())
                .createdAt(exercise.getCreatedAt())
                .updatedAt(exercise.getUpdatedAt())
                .build();
    }
}

File: hyperlift--frontend/src/services/exerciseService.ts
import api from './axiosConfig';
import { ExerciseRequest, ExerciseResponse, ApiResponse } from '../types/AppTypes';

const EXERCISE_URL = '/exercises';

export const exerciseService = {
  getAllExercises: async (): Promise<ApiResponse<ExerciseResponse[]>> => {
    const response = await api.get(EXERCISE_URL);
    return response.data;
  },

  getExerciseById: async (id: number): Promise<ApiResponse<ExerciseResponse>> => {
    const response = await api.get(`${EXERCISE_URL}/${id}`);
    return response.data;
  },

  getExercisesByMuscleGroup: async (muscleGroup: string): Promise<ApiResponse<ExerciseResponse[]>> => {
    const response = await api.get(`${EXERCISE_URL}/muscle-group/${muscleGroup}`);
    return response.data;
  },

  createExercise: async (data: ExerciseRequest): Promise<ApiResponse<ExerciseResponse>> => {
    const response = await api.post(EXERCISE_URL, data);
    return response.data;
  },

  updateExercise: async (id: number, data: ExerciseRequest): Promise<ApiResponse<ExerciseResponse>> => {
    const response = await api.put(`${EXERCISE_URL}/${id}`, data);
    return response.data;
  },

  deleteExercise: async (id: number): Promise<ApiResponse<void>> => {
    const response = await api.delete(`${EXERCISE_URL}/${id}`);
    return response.data;
  },
};

export default exerciseService;
